### DISCUSSION TITLE
Disk IO only shown for a few disks

### DISCUSSION THREAD

**Discussion by MPeti1** [aristocratos/bpytop#287] [Q&A] [answered]
The disk section of the memory graph has a list of disks, showing their names and capacity in percentage and bytes.
Currently I have 4 disks there (partitions, actually) (I've the only physical setting turned on): the rootfs, the boot partition (both of these on the same device), swap, and a data partition on an other device.
I have the "Show io stat" setting turned on in the options menu, but the IO stat line has only appeared for 2 devices from the 4: the boot partition and the data partition on the other device.

How could I make the IO meter work for the rootfs too?

I checked my fstab to find if rootfs is unique in some aspects, and it's mounted with the `noatime` option beside the `defaults`.
<details><summary>Is it possible that this option is the culprit?</summary>
I wanted to check this myself earlier, but I haven't found the right moment to reboot the machine, because it plays an important role in the local network.</details>
If so, can I do something to keep noatime, but also have an IO meter for this partition?

**Comment by aristocratos** [comment, marked as answer]
The way the IO stats work in bpytop is by using the two `psutil` functions `psutil.disk_partitions()` (to get the partitions stats) and `psutil.disk_io_counters(perdisk=True)` (to get the IO stats).

But `psutil.disk_io_counters` does not return the full path to the device, so bpytop will try to match the device name to a partition, this  can fail in the case of weird device names, like for raided disks or encrypted partitions.

I don't have a solution for this happening right now, but if anyone have any ideas I'm all ears.

**Reply by MPeti1** [reply]
Aha!

`psutil.disk_io_counters(perdisk=True)` alongside a bunch of ram and loop devices, gives me sdb and sdc and their numbered versions for partitions. only device names (truncated to simple name instead of devtmpfs path) and stats are here.

`psutil.disk_partitions()` returns the mounted partitions, with devices of sdb1, sdc2 (these are being matched by bpytop and shown in the menu) but the device of the rootfs is just /dev/root instead of the disk's name in /dev (why though? it's raspbian on an RPi 4)

I was thinking on a fix, and I had a few ideas.
- /etc/fstab is world readable by default on raspbian, not sure about other linux os-es though
- if the system has `lsblk`, it can provide you the information you need
- I was thinking about that /dev/root too, especially if it's a symlink to /dev/sdb2 (which the rootfs partition actually is) or what it is actully, and the thing is.. that it doesn't exist. I've included a screenshot below about what psutil says about disk partitions.

![image](https://user-images.githubusercontent.com/9215491/114463016-62a02800-9bd3-11eb-97b6-39c13f1e7cad.png)

**Reply by MPeti1** [reply]
I think for my particular it case may be enough to find out why psutil refers to the rootfs as having the /dev/root device,
but for other cases.. only lsblk comes to my mind (I'm not an expert linux admin, though!), but since it's probably not available on every system, I see 2 options (which includes lsblk, that is):
- make it an optional feature, detect if lsblk is available and if not, use the old beahvior, otherwise get devcie paths from there
- a python library that implements the funcionality of lsblk (if one exists) could be an option too.
There are already optional dependencies for OSX, so if this does not complicate it too much then it could work, I think.

Sample output of lsblk:
![image](https://user-images.githubusercontent.com/9215491/114464033-c119d600-9bd4-11eb-9ac2-075bd7ce9075.png)

Sample output of lsblk with my preferred options:
![image](https://user-images.githubusercontent.com/9215491/114464173-edcded80-9bd4-11eb-894e-68f429ca63e3.png)
options: `lsblk --output MODEL,NAME,LABEL,PATH,FSSIZE,FSTYPE,MOUNTPOINT,PARTUUID`

**Reply by aristocratos** [reply]
That's a great idea! 
`lsblk` should be available on most Linux systems, so could simply add a function that parses `lsblk` output (if available) and maps all partition names to the correct "master" /dev device. It would only need to be run at start and if a change in number of partitions is detected. Should greatly increase the chance of finding correct IO for all partitions.

Will take a look at it when I've got time.
